generals/gui/event_handler.py

```diff
@@ -1,77 +1,179 @@
+L = pygame.K_l
+
+
+class Command:
+    def __init__(self):
+        self.quit: bool = False
+
+
+class ReplayCommand(Command):
+    def __init__(self):
+        super().__init__()
+        self.frame_change: int = 0
+        self.speed_change: float = 1.0
+        self.restart: bool = False
+        self.pause: bool = False
+
+
+class GameCommand(Command):
+    def __init__(self):
+        super().__init__()
+        raise NotImplementedError
+
+
+class TrainCommand(Command):
+    def __init__(self):
+        super().__init__()
 
 
 class EventHandler:
-    def __init__(self, properties: Properties, from_replay=False):
+    def __init__(self, properties: Properties):
         """
         Initialize the event handler.
 
         Args:
             properties: the Properties object
-            from_replay: bool, whether the game is from a replay
         """
         self.properties = properties
-        self.from_replay = from_replay
+        self.mode = properties.mode
+        self.handler_fn = self.initialize_handler_fn()
+        self.command = self.initialize_command()
+
+    def initialize_handler_fn(self):
+        """
+        Initialize the handler function based on the mode.
+        """
+        if self.mode == "replay":
+            return self.__handle_replay_key_controls
+        elif self.mode == "game":
+            return self.__handle_game_key_controls
+        elif self.mode == "train":
+            return self.__handle_train_key_controls
+        raise ValueError("Invalid mode")
 
-    def handle_events(self):
+    def initialize_command(self):
+        """
+        Initialize the command type based on the mode.
+        """
+        if self.mode == "replay":
+            return ReplayCommand
+        elif self.mode == "game":
+            return GameCommand
+        elif self.mode == "train":
+            return TrainCommand
+        raise ValueError("Invalid mode")
+
+    def handle_events(self) -> Command:
         """
         Handle pygame GUI events
         """
-        control_events = {
-            "time_change": 0,
-        }
+        command = self.command()
         for event in pygame.event.get():
-            if event.type == pygame.QUIT or (
-                event.type == pygame.KEYDOWN and event.key == pygame.K_q
-            ):
-                pygame.quit()
-                quit()
-
-            if event.type == pygame.KEYDOWN and self.from_replay:
-                self.__handle_key_controls(event, control_events)
+            if event.type == pygame.QUIT:
+                command.quit = True
+            if event.type == pygame.KEYDOWN:
+                command = self.handler_fn(event, command)
             elif event.type == pygame.MOUSEBUTTONDOWN:
                 self.__handle_mouse_click()
+        return command
 
-        return control_events
-
-
-    def __handle_key_controls(self, event, control_events):
+    def __handle_replay_key_controls(self, event: Event, command: Command) -> Command:
```

Comment:
why the `command` parameter and returning? why not just use `self.command`?

Reply:
will be resolved with the proposed strategy pattern